### PR TITLE
Package ocaml-monadic.0.4.0

### PR DESCRIPTION
**Pull request by zepalmer**
### `ocaml-monadic.0.4.0`

OCaml monadic syntax extensions

A simple PPX language extension to include monadic operations that
blend nicely with OCaml syntax.

---
* Homepage: http://github.com/zepalmer/ocaml-monadic
* Source repo: https://github.com/zepalmer/ocaml-monadic.git
* Bug tracker: https://github.com/zepalmer/ocaml-monadic/issues

---

:camel: Pull-request generated by opam-publish v0.3.5

### CODE REVIEW

**Comment by mseri** [comment]
The CI has these failures for ocaml 4.02 and 4.03:

```
+ jbuilder "runtest" "-p" "ocaml-monadic" "-j" "4" (CWD=/home/travis/.opam/system/build/ocaml-monadic.0.4.0)
-          ppx src_test/test.pp.ml (exit 1)
- (cd _build/default && ./.ppx/ocaml-monadic/ppx.exe --dump-ast -o src_test/test.pp.ml --impl src_test/test.ml)
- File "src_test/test.ml", line 92, characters 9-10:
- Error: Syntax error
```

**Comment by kit-ty-kate** [comment]
As you seem to be the maintainer of `jhubllib` you might help here. It seems that the archives of `jhubllib` have a different checksum than originally. Could you update their checksum in this PR as well ?

The error message was not being really helpful, I'm sorry. I reported this to https://github.com/ocaml/opam/issues/3393

**Comment by zepalmer** [comment]
I'd be happy to try to correct the other problems, but I'm afraid I don't understand them.  I use `opam-publish` in order to publish packages, so I've never had to delve into this kind of detail before.  There are a couple things I don't understand.

1. I cloned `ocaml/opam-repository` myself and checked out to the branch that `opam-publish` apparently created for me.  I then looked at the `url` file for my PR, which contains

   ```
   http:
     "https://github.com/JHU-PL-Lab/jhu-pl-lib/archive/da46a9e849d8d8f627e59025726d90e64268c9f1.zip"
   checksum: "e8284912f1eaadd88c5327103893370d"
   ```

   This checksum was previously automatically generated for me by `opam-publish`.  Based on its length, I'm guessing it's an MD5; is that right?  I can verify that `md5sum` on the `.zip` does not return the same results as recorded in that file.  If so, would it be enough for me to change the checksum in that file?

2. I don't believe anyone has force-pushed to that repository or done anything else invasive like that.  What could've caused this scenario to arise?

**Comment by zepalmer** [comment]
I attempted the adjustment that I described above and reverted it.  If I change the checksums that appear in the "url" files of those three revdeps, everything fails much more aggressively (10/11 failed checks).  Any advice?

**Comment by mseri** [comment]
I think this could be merged. If you figure out a fix for jhubllib it could go in a separate PR

File: packages/jhupllib/jhupllib.0.1.1/url [hunk 1]
```diff
@@ -1,3 +1,3 @@
 http:
   "https://github.com/JHU-PL-Lab/jhu-pl-lib/archive/824006ad7c0b05d015c3449f168bb5e99882bee5.zip"
-checksum: "4e5d5216bc467bdfbf561966d4a748b3"
+checksum: "88cf95af4ba9053d0873577953f714be"
```
Comment (mseri) [inline packages/jhupllib/jhupllib.0.1.1/url]: My `md5sum` for that archive is `3996f5c6a55b8bd37db7c3db11724641`. How did you get that value?

Also, `http:` as field is deprecated, please use `archive:`

**Comment by kit-ty-kate** [comment]
It seems that pds-reachability.0.2.1 needs some constraints on one of its constraints:
```
#=== ERROR while compiling pds-reachability.0.2.1 =============================#
# context              2.0.0~rc | linux/x86_64 | ocaml-base-compiler.4.06.1 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.06/.opam-switch/build/pds-reachability.0.2.1
# command              ~/.opam/4.06/bin/ocaml setup.ml -build
# exit-code            1
# env-file             ~/.opam/log/pds-reachability-18-b36dbe.env
# output-file          ~/.opam/log/pds-reachability-18-b36dbe.out
### output ###
# ocamlfind ocamlopt unix.cmxa -I /home/opam/.opam/4.06/lib/ocamlbuild /home/opam/.opam/4.06/lib/ocamlbuild/ocamlbuildlib.cmxa -linkpkg myocamlbuild.ml /home/opam/.opam/4.06/lib/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_analysis.ml > src/pds-reachability/pds_reachability_analysis.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_basis.ml > src/pds-reachability/pds_reachability_basis.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_utils.ml > src/pds-reachability/pds_reachability_utils.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_utils.cmo src/pds-reachability/pds_reachability_utils.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_structure.ml > src/pds-reachability/pds_reachability_structure.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_basis.cmo src/pds-reachability/pds_reachability_basis.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_types.ml > src/pds-reachability/pds_reachability_types.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_types_stack.ml > src/pds-reachability/pds_reachability_types_stack.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -w -37 -I src/pds-reachability -o src/pds-reachability/pds_reachability_types_stack.cmo src/pds-reachability/pds_reachability_types_stack.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_types.cmo src/pds-reachability/pds_reachability_types.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_work.ml > src/pds-reachability/pds_reachability_work.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_work_collection.ml > src/pds-reachability/pds_reachability_work_collection.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_work.cmo src/pds-reachability/pds_reachability_work.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_structure.cmo src/pds-reachability/pds_reachability_structure.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_work_collection.cmo src/pds-reachability/pds_reachability_work_collection.ml
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability_work_collection_templates.ml > src/pds-reachability/pds_reachability_work_collection_templates.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamldep -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -modules src/pds-reachability/pds_reachability.ml > src/pds-reachability/pds_reachability.ml.depends
# /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_analysis.cmo src/pds-reachability/pds_reachability_analysis.ml
# + /home/opam/.opam/4.06/bin/ocamlfind ocamlc -c -g -annot -bin-annot -package batteries -package jhupllib -package ocaml-monadic -package ppx_deriving.std -package ppx_deriving_yojson -package threads -package yojson -thread -I src/pds-reachability -o src/pds-reachability/pds_reachability_analysis.cmo src/pds-reachability/pds_reachability_analysis.ml
# File "src/pds-reachability/pds_reachability_analysis.ml", line 8, characters 5-25:
# Error: Unbound module Eager_nondeterminism
# Command exited with code 2.
# E: Failure("Command ''/home/opam/.opam/4.06/bin/ocamlbuild' src/pds-reachability/pds-reachability.cma src/pds-reachability/pds-reachability.cmxa src/pds-reachability/pds-reachability.a src/pds-reachability/pds-reachability.cmxs test/test.native -use-ocamlfind -tag debug' terminated with error code 10")
```

**Comment by mseri** [comment]
Looks much better, thanks. From a brief look at the repos, I think the error that we see now is due to pds-reachability getting jhupllib 0.1 instead of 0.1.1. There is probably a lower bound on the dependencies needed

**Comment by mseri** [comment]
@kit-ty-kate you have beaten me on the timing :D

**Comment by zepalmer** [comment]
Thanks for the help on this.  I think I just sorted it; I've pushed and we'll see how the tests go.  I suppose the tests were different when I pushed pds-reachability 0.2.1; I can't imagine how else it would've gotten into this state.

I'll update the other metadata in my repositories once I see all of this work out.  :)  Thanks again for the knowledge and support!

**Comment by mseri** [comment]
Thanks! The two failures in the CI are independent of your changes. If Travis is happy we can merge.

**Comment by zepalmer** [comment]
Travis does not appear to be happy, but it's a 404 trying to download ppx_import.

**Comment by hcarty** [comment]
I think the 404 is being addressed in #12137.

**Comment by zepalmer** [comment]
@hcarty I think you might be right.  How does one restart the Travis CI check?  It's the only one left, but it doesn't have the same interface as the OCaml Labs CI.

**Comment by hcarty** [comment]
@mseri or @kit-ty-kate Can you restart the Travis check?  Everything else looks green now.

**Comment by kit-ty-kate** [comment]
All green. Thanks a lot !!

**Pull request merged**